subsys/testsuite/include/interrupt_util.h

```diff
@@ -101,30 +101,47 @@ static inline void trigger_irq(int irq)
 }
 
 #elif defined(CONFIG_X86)
-#include <drivers/interrupt_controller/loapic.h>
-#include <sys/arch_interface.h>
 
 #define TRIGGER_IRQ_INT(vector) __asm__ volatile("int %0" : : "i" (vector) : "memory")
 
 /*
- * Write Local APIC's ICR to trigger IPI for testing
+ * We can emulate the interrupt by sending the IPI to
+ * core itself by the LOAPIC for x86 platform.
+ *
+ * In APIC mode, Write LOAPIC's ICR to trigger IPI,
+ * the LOAPIC_ICR_IPI_TEST  0x00004000U means:
  * Delivery Mode: Fixed
  * Destination Mode: Physical
  * Level: Assert
  * Trigger Mode: Edge
  * Destination Shorthand: No Shorthand
  * Destination: depends on cpu_id
+ *
+ * In X2APIC mode, this no longer works. We emulate the
+ * interrupt the by writing the IA32_X2APIC_SELF_IPI MSR
+ * to send IPI to the core itself via LOAPIC also.
+ * According to SDM vol.3 chapter 10.12.11, the bit[7:0]
+ * for setting the vector is only needed.
  */
-#define LOAPIC_ICR_IPI_TEST  0x00004000U
-
 static inline void trigger_irq(int vector)
 {
+#ifdef CONFIG_X2APIC
+#include <drivers/interrupt_controller/loapic.h>
+
+#define VECTOR_MASK 0xFF
+
+	x86_write_x2apic(LOAPIC_SELF_IPI, ((VECTOR_MASK & vector)));
+#else
+#include <sys/arch_interface.h>
+#define LOAPIC_ICR_IPI_TEST  0x00004000U
+
```

Comment:
include headers and define macros, out of the function.

Reply:
Ok, move it out of the function. Thanks!